File: 1289-minimum-falling-path-sum-ii/1289-minimum-falling-path-sum-ii.py
# No TLE
# Space optimization + heap
class Solution:
    def minFallingPathSum(self, grid: List[List[int]]) -> int:
        l = len(grid)
        prev, maxhp = [grid[-1][i] for i in range(l)], []
        for i in prev:
            heappush(maxhp, -i)
            if len(maxhp)>2:
                heappop(maxhp)
        
        for row in range(l-2, -1, -1):
            curr = [0]*l
            maxhp2 = []
            for col in range(l):
                curr[col] = -maxhp[1] if -maxhp[1] != prev[col] else -maxhp[0]
                curr[col] += grid[row][col]    
                
                heappush(maxhp2, -curr[col])
                if len(maxhp2)>2:
                    heappop(maxhp2)
            
            prev = curr
            maxhp = maxhp2
        
        return min(prev)


# TLE
# tabulation soln without using heap

# class Solution:
#     def minFallingPathSum(self, matrix: List[List[int]]) -> int:
#         l=len(matrix)
#         dp = [[10**6]*l for _ in range(l)]
        
#         for row in range(l-1, -1, -1):
#             for col in range(l):
#                 if row != l-1:
#                     for i in range(l):
#                         if i!=col:
#                             dp[row][col] = min(dp[row][col], dp[row+1][i])
#                     dp[row][col] += matrix[row][col]
#                 else:
#                     dp[row][col] = matrix[row][col]
        
#         return min(dp[0])

File: 1289-minimum-falling-path-sum-ii/1289-minimum-falling-path-sum-ii.java

class Solution {
    public int minFallingPathSum(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        int[][] dp = new int[m][n];
        dp[m-1] = grid[m-1];

        for(int i=m-2; i>=0; --i){
            for(int j=0; j<n; ++j){
                dp[i][j] = 1000_000_000;
                for(int x=0; x<m; ++x){
                    if(x!=j) dp[i][j] = Math.min(dp[i][j], dp[i+1][x]);
                }
                dp[i][j] += grid[i][j];
            }
        }
        
        int ans = Integer.MAX_VALUE;
        for(int d: dp[0]) ans = Math.min(ans, d);
        return ans;
    }
}






// // mis-understood the question
//
// class Solution {
//     Integer[][] dp;
//     public int minFallingPathSum(int[][] grid) {
//         dp = new Integer[grid.length][grid[0].length];
//         int ans = Integer.MAX_VALUE;
//         for(int i=0; i<grid[0].length; ++i)
//             ans = Math.min(ans, dfs(0, i, grid));
        
//         return ans;
//     }
//     int dfs(int i, int j, int[][] grid){
//         if(i==grid.length) return 0;
        
//         if(dp[i][j] != null) return dp[i][j];
        
//         int ans = 1000000000;
//         if(j>0) ans = Math.min(ans, dfs(i+1, j-1, grid)+grid[i][j]);
//         if(j<grid[0].length-1) ans = Math.min(ans, dfs(i+1, j+1, grid)+grid[i][j]);
//         if(ans >= 1000000000) ans = grid[i][j];
//         // ans += grid[i][j];
        
//         dp[i][j] = ans;
        
//         return ans;
//     }
// }